src/api/TKRecoveryManager.m

```diff
@@ -0,0 +1,189 @@
+          onSuccess:^(NSString *verificationId_) {
+              verificationId = verificationId_;
+              
+              //Generate keys for later recovery process
+              crypto = [[TKCrypto alloc] initWithEngine:[cryptoEngineFactory createEngine:memberId]];
+              [crypto generateKeys];
+              
+              onSuccess();
+          }
+          onError:onError];
+     }
+     onError:onError];
+}
+
+- (void)verifyRecoveryCode:(NSString *)code
+                 onSuccess:(OnSuccessWithBoolean)onSuccess
+                   onError:(OnError)onError {
+    
+    if (crypto == nil) {
+        onError([NSError
+                 errorFromErrorCode:kTKErrorInvalidRecoveryProcess
+                 details:@"Please call beginRecovery before verifyRecoveryCode"]
+                );
+    }
+    
+    Key *key = [crypto getKeyInfo:Key_Level_Privileged
+                           reason:TKLocalizedString(
+                                    @"Signature_Reason_RecoverMember",
+                                    @"Approve to recover a Token member account")
+                          onError:onError];
+    if (!key) {
+        return;
+    }
+    [unauthenticatedClient
+     getRecoveryOperation:verificationId
+     code:code
+     privilegedKey:key
+     onSuccess:^(MemberRecoveryOperation *op) {
+         recoveryOperation = op;
+         onSuccess(true);
+     }
+     onError:onError];
+}
+
+- (void)completeRecovery:(OnSuccessWithTKMember)onSuccess
+                 onError:(OnError)onError {
+    
+    if (crypto == nil) {
+        onError([NSError
+                 errorFromErrorCode:kTKErrorInvalidRecoveryProcess
+                 details:@"Please call beginRecovery before completeRecovery"]
+                );
+    }
+    
+    if (recoveryOperation == nil) {
+        onError([NSError
+                 errorFromErrorCode:kTKErrorInvalidRecoveryProcess
+                 details:@"Please call verifyRecoveryCode before completeRecovery"]
+                );
+    }
+    
+    NSString *reason = TKLocalizedString(
+                         @"Signature_Reason_RecoverMember",
+                         @"Approve to recover a Token member account");
+    NSArray<Key *> *keys = [crypto getKeyInfos:reason
+                                       onError:onError];
+    if (!keys) {
+        return;
+    }
+    
+    NSMutableArray<MemberOperation *> *operations = [NSMutableArray array];
+    MemberOperation *recoverOp = [MemberOperation message];
+    recoverOp.recover = recoveryOperation;
+    [operations addObject:recoverOp];
+
+    for (Key *key in keys) {
+        MemberOperation *addKey = [MemberOperation message];
+        addKey.addKey.key = key;
+        [operations addObject:addKey];
+    }
+
+    [unauthenticatedClient
+     updateMember:memberId
+     crypto:crypto
+     operations:operations
+     metadataArray:[NSArray array]
+     reason:reason
+     onSuccess:^(Member *member){
+         [unauthenticatedClient
```

Comment:
I think we need to add retry logic to this call because of the possible problem with the last step not being executed. Perhaps that's another ticket for the whole sdk

Reply:
Or at least make sure that it's idempotent. If client sends an updateMember call with the same last hash and contents, the server will respond with the same message. If the app crashes between these two calls and the developer retries, we need to either structure this differently or make sure that the last hash doesn't get updated (otherwise the updateMember will be rejected and the verify call will not be called)